### PR TITLE
config: simplify most of the config handling and fix some bugs

### PR DESCRIPTION
**Pull request by c00w**
This PR combines a number of cleanups in one PR. If any of the specific cleanups don't seem to make sense, let me know and I can remove them.

Cleanups

- This PR adds a set of test suites for the config module code, which handles basically all the APIs and ways it is used. Please let me know if you see anything critical that is not tested that I missed. This test suite is primarily used as the regression test suite for later changes in this diff. Note that there is some dynamo specific testing of the config module, but it isn't as verbose.
- I removed all internal usage of shallow_copy_dict. Those usages could all use the deep copy, and did not depend on the reference behavior of certain config values that shallow_copy_dict allows.
- I removed shallow copy semantics for configuration with a deprecation warning. I think this requires a release note, so hopefully I did that correctly. Let me know if we want to continue to expose shallow copy value semantics, but I just can't find a case where I expect anyone would want it. It also complicated later internal changes to the API (i.e. breaking apart various layers of the config changes).
- I fixed what I believe is a bug in how hashes are calculated on configs. In particular, if you got the hash, then made a config change, and then got the hash again, it would not update the hash. @oulgen, please let me know if I'm misunderstanding this behavior and it is desired.
- I switched our multiple implementations of iterating through the dictionary to a single one. This is primarily to make later changes easier, but it also makes it clear how inconsistent our various config ignoring options are. Let me know if people would be interested in me unifying the various options for ignoring config values.
- I updated the test patcher (not the performance critical one, just the normal one), to use __setattr__ and __getattr__ to remove direct API access to the underlying config fetcher.

For release notes, Not sure exactly how to communicate this, but something like
"ConfigModule.to_dict, and ConfigModule.shallow_copy_dict no longer retain their shallow copy semantics, which allowed reference values objects to be modified. If you wish to modify the config object, call load_config explicitly".

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @rec

### CODE REVIEW

File: torch/utils/_config_module.py [hunk 1]
```diff
@@ -153,69 +154,84 @@ def __getattr__(self, name: str) -> Any:
...
         # must support delete because unittest.mock.patch deletes
         # then recreate things
         del self._config[name]
 
+    def _get_dict(
+        self,
+        ignored_keys: Optional[List[str]] = None,
+        ignored_prefixes: Optional[List[str]] = None,
+        skip_default: bool = False,
+    ) -> Dict[str, Any]:
```
Comment (ezyang) [inline torch/utils/_config_module.py:167]: doc?  At least, it's not immediately obvious to me reading the method name what this does

Comment (ezyang) [reply torch/utils/_config_module.py:167]: Factored out from `_codegen_config` it seems

File: torch/utils/_config_module.py [hunk 2]
```diff
@@ -153,69 +154,84 @@ def __getattr__(self, name: str) -> Any:
...
+        return config
+
     def save_config(self) -> bytes:
         """Convert config to a pickled blob"""
-        config = dict(self._config)
-        for key in config.get("_save_config_ignore", ()):
-            config.pop(key)
-        return pickle.dumps(config, protocol=2)
+        return pickle.dumps(
+            self._get_dict(ignored_keys=self._config.get("_save_config_ignore", ())),
```
Comment (ezyang) [inline torch/utils/_config_module.py:187]: It occurs to me that now that we're supporting field() style definitions of config values, we don't have to defines these out of band anymore, we could just have them be magic objects that are processed on load and store them separately. 🤔

**Review by ezyang — APPROVED** [review]
Thanks, nice work.

**Review by jansel — APPROVED** [review]

**Review by jovianjaison — APPROVED** [review]

File: torch/utils/_config_module.py (hunk 2 above)
Comment (c00w) [reply torch/utils/_config_module.py:187]: i.e. foo = Config(value=True, skip_save_config=True)?

Happy to send a PR :).

File: torch/utils/_config_module.py (hunk 1 above)
Comment (c00w) [reply torch/utils/_config_module.py:167]: Updated the docstring :).

**Comment by c00w** [comment]
Failing job appears unrelated to this diff - and I belive is failing on hud - https://github.com/pytorch/pytorch/actions/runs/11446310391/job/31846657178 -

**Pull request closed without merging**